### PR TITLE
[PLU-790-1]: add attachment support for SES

### PR DESCRIPTION
**Pull request by m0nggh**
### TL;DR

Adds attachment support to the SES email path, gated behind a new `ses_attachments_enabled` LaunchDarkly flag.

### What changed?

- Introduced a `buildRawEmail` helper (`build-raw-mime.ts`) that uses nodemailer's `MailComposer` to construct a complete RFC 5322 MIME message. This handles RFC 2047 encoded-words for non-ASCII subjects and filenames, base64 line wrapping, and MIME boundaries correctly.
- When an email has attachments and both `ses_enabled` and `ses_attachments_enabled` are on for all recipients, the SES path now sends a raw MIME message via `Content.Raw` instead of `Content.Simple`. Emails without attachments continue to use `Content.Simple` unchanged.
- Added a 10MB total attachment size guard (`AttachmentSizeExceededError`) on the SES path to match Postman's server-side limit. This is enforced before the MIME message is built and maps to `ATTACHMENT-SIZE-EXCEEDED` in `getSesErrorStatus`.
- Added `isSesAttachmentsEnabledForRecipient` in `ses-email-helper.ts`, which checks the `ses_attachments_enabled` LaunchDarkly flag per recipient. Attachments only go via SES when this flag is enabled for every recipient in addition to `ses_enabled`.
- The send log now includes `attachmentCount` and `totalAttachmentBytes` for rollout visibility.
- Added `nodemailer` (`9.0.3`) and `@types/nodemailer` (`8.0.1`) as dependencies.

### How to test?

- Enable both `ses_enabled` and `ses_attachments_enabled` in LaunchDarkly for a test recipient and send an email with attachments — verify it is delivered via SES as a raw MIME message and contains the `X-Plumber-Transport: ses` header.
- Enable `ses_enabled` but leave `ses_attachments_enabled` off — verify that an email with attachments falls back to Postman.
- Send an email with attachments totalling more than 10MB on the SES path — verify it is rejected with `ATTACHMENT-SIZE-EXCEEDED` before any SES API call is made.
- Run the new unit tests: `build-raw-mime.test.ts` and the updated `send-transactional-email.test.ts`.

### Why make this change?

The SES direct-send path previously fell back to Postman for any email containing attachments. This change extends SES support to cover attachments, allowing a full migration away from Postman for transactional emails while preserving the existing suppression checks, configuration set, metrics, and error-handling behaviour.

### CODE REVIEW

**Comment by m0nggh** [comment]
* **#1809** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1809?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1808** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1808?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#1807** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1807?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1807?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#1806** <a href="https://app.graphite.com/github/pr/opengovsg/plumber/1806?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `feat/ses-attachments/trunk`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Comment by kevinkim-ogp** [comment]
## Code Review

**Mode:** Light (diff-only), extended to heavy for the core files (`email-helper.ts`, `ses-email-helper.ts`, `build-raw-mime.ts`) plus the caller (`send-transactional-email/index.ts`) and `parameters.ts`, given attachment/email-sending changes touch header-construction code.

Ran the new/updated unit tests locally (`build-raw-mime.test.ts`, `send-transactional-email.test.ts`) — all 37 pass. I also checked nodemailer's header-building code for CRLF/header-injection risk given that `subject`, `from`, `to`, `cc`, `replyTo`, and attachment filenames are user-supplied flow-step values — nodemailer strips `\r?\n|\r` from header values before emitting them, so this isn't exploitable here.

### Findings

#### 1. 🟡 Medium · ⚡ Performance
**Where:** [`email-helper.ts:193-203`](https://github.com/opengovsg/plumber/blob/feat/ses-attachments/send-attachments/packages/backend/src/apps/postman/common/email-helper.ts#L193-L203) (called from the per-recipient loop at `email-helper.ts:324-329`)

**Issue:** `buildRawEmail` (which base64-encodes the full attachment set) is invoked once per recipient inside `activeRecipients.map`, even though attachments, subject, body, cc, and replyTo are identical across recipients — only the `To` address differs. Since `destinationEmail` supports comma-separated batches (up to ~50 recipients per the comment in `parameters.ts`) and attachments can now be up to the new 10MB parity cap, a single flow execution could re-run full base64 MIME construction up to ~50 times — e.g. ~50×10MB of base64 work — where the previous Postman path only appended raw buffers into `FormData` per recipient (no re-encoding). This is a real step up in worker CPU/memory cost per send that didn't exist before this change.

**Fix:** Consider building the raw MIME message once per email and only varying the SES `Destination.ToAddresses` per call, rather than rebuilding per recipient.

### Summary

**Verdict:** Comment (non-blocking)

**Done well:** The size guard runs before the more expensive MIME build, error-name-based mapping in `getSesErrorStatus` cleanly reuses the existing Postman status vocabulary, and the new unit tests cover the interesting encoding edge cases (non-ASCII subject/filename, multi-attachment, size guard) rather than just the happy path.

**Comment by kevinkim-ogp** [comment]
### Code review

No issues found. Checked for bugs and CLAUDE.md compliance.

🤖 Generated with [Claude Code](https://claude.ai/code)

<sub>- If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

**Comment by kevinkim-ogp** [comment]
Following up on the [performance finding](https://github.com/opengovsg/plumber/pull/1807#issuecomment-4913102029) above — here's a concrete way to hoist `buildRawEmail` out of the per-recipient loop without losing per-recipient correctness (the `To:` header still needs to reflect the actual recipient, since mail clients display it and it shouldn't leak other recipients' addresses).

The key insight: `MailComposer` only emits a `To:` header if `mail.to` is set (`mail-composer/index.js`: `if (this.mail[key]) this.message.setHeader(...)`), so the shared/expensive part (attachments, subject, body, cc, reply-to) can be built once without `to`, and the cheap per-recipient part (`To:`) can be prepended as raw text afterward — safe because the header block isn't terminated until the first blank line, and `buildHeaders()` doesn't emit a leading blank line itself.

**`build-raw-mime.ts`** — make `to` optional (keep it supported for direct/simple callers and existing unit tests) and add a cheap post-processing helper:

```ts
export interface RawEmailInput {
  from: string
  to?: string // omit to build a recipient-agnostic message; combine with withRecipient()
  cc?: string[]
  replyTo?: string
  subject: string
  html: string
  attachments: { fileName: string; data: Uint8Array }[]
  headers?: Record<string, string>
}

// ...inside buildRawEmail's composer options:
...(input.to && { to: input.to }),

// Cheap per-recipient step: prepend a `To:` header line to an already-built
// message. Safe because MailComposer's header block isn't terminated until
// the first blank line, and buildHeaders() never emits a leading blank line.
export function withRecipient(rawMessage: Buffer, to: string): Buffer {
  return Buffer.concat([Buffer.from(`To: ${to}\r\n`, 'utf-8'), rawMessage])
}
```

**`email-helper.ts`** — build the shared message once in `sendTransactionalEmails`, before the `activeRecipients.map` loop, and re-throw a captured size-guard error inside each recipient's existing try/catch so `getSesErrorStatus`/`dataOut` status-mapping is unchanged:

```ts
let sharedRawMessage: Buffer | undefined
let attachmentBuildError: unknown

if (useSes && email.attachments?.length && activeRecipients.length) {
  const totalAttachmentBytes = email.attachments.reduce((s, a) => s + a.data.byteLength, 0)
  if (totalAttachmentBytes > SES_MAX_TOTAL_ATTACHMENT_SIZE) {
    attachmentBuildError = new AttachmentSizeExceededError()
  } else {
    sharedRawMessage = await buildRawEmail({
      from: formatFromAddress(email.senderName, appConfig.ses.fromAddress),
      cc: ccAddressesToSend,
      replyTo: email.replyTo,
      subject: email.subject,
      html: sanitizeEmailHtml(email.body),
      attachments: email.attachments,
      headers: { 'X-Plumber-Transport': 'ses' },
    })
  }
}

// inside the per-recipient map, in place of the current `sendViaSes(...)` call:
if (useSes) {
  if (attachmentBuildError) throw attachmentBuildError
  return await sendViaSes(recipientEmail, email, ccAddressesToSend, sharedRawMessage)
}
```

`sendViaSes` then shrinks to just `Content: { Raw: { Data: withRecipient(sharedRawMessage, recipientEmail) } }` for the attachment branch — no more per-recipient size check or `buildRawEmail` call.

Net effect: base64-encoding the attachment set happens once per email instead of once per recipient (up to ~50x today), while each recipient still gets a correct, individual `To:` header. Existing tests should hold as-is since `to` stays optional/backward-compatible and the size-guard/status-mapping path is unchanged, just relocated.

**Review by kevinkim-ogp — APPROVED** [review]
lgtm!  
maybe check out claude's finding about invoking `buildRawEmail` per recipient. can assess if its worth fixing

File: packages/backend/src/apps/postman/common/email-helper.ts [hunk 1]
```diff
@@ -135,6 +137,21 @@ async function sendViaPostman(
...
+// SES path bypasses. Compared against raw (pre-base64) bytes — the same basis
+// Postman uses. SES itself allows up to 40MB, so this is the parity cap, not an
+// SES limit.
+const SES_MAX_TOTAL_ATTACHMENT_SIZE = 10 * 1024 * 1024
+
+// Thrown by the SES path when attachments exceed the parity cap. Mapped to
+// ATTACHMENT-SIZE-EXCEEDED by getSesErrorStatus (matched by name).
+class AttachmentSizeExceededError extends Error {
+  constructor() {
+    super('Total attachment size exceeds 10MB')
```
[the remark this answers is not in the document]
Comment (m0nggh) [reply packages/backend/src/apps/postman/common/email-helper.ts:150]: doesnt happen cus its used for internal logging only
